### PR TITLE
Add formatting check to CI pipeline

### PR DESCRIPTION
**Pull request by Vexu**
Reopening of #8219 now that the parsing issue has been fixed. I also added a `check-ast` on the self hosted compiler `main.zig` since that should also test the standard library.

### CODE REVIEW

**Comment by g-w1** [comment]
Ast-check does not follow imports, so you should do `find .. -name "*.zig" | xargs ./zig ast-check` or something similar.

File: ci/azure/linux_script [hunk 1]
```diff
@@ -59,6 +59,10 @@ unset CXX
 
 make $JOBS install
 
+# look for formatting errors
+release/bin/zig fmt --check ../lib/std ../doc ../src ../tools ../test ../build.zig
+release/bin/zig ast-check ../src/main.zig
```
Comment (andrewrk) [inline ci/azure/linux_script:64]: I suggest we only do the zig fmt check for now, for a few reasons:
 * This doesn't actually check any files other than src/main.zig
 * std lib AST checking is already happening with every stage2 test
 * It's an unrelated change to adding zig fmt checking 
 * we may want to consider taking advantage of the thread pool and cache, which is already happening with std lib AST checking

File: ci/azure/linux_script [hunk 2]
```diff
@@ -59,6 +59,10 @@ unset CXX
 
 make $JOBS install
 
+# look for formatting errors
+release/bin/zig fmt --check ../lib/std ../doc ../src ../tools ../test ../build.zig
```
Comment (andrewrk) [inline ci/azure/linux_script:63]: Suggestion, to make it more clear what to do for a contributor whose PR is not passing:

` || echo "Please run 'zig fmt' to fix the non-conforming files listed above."`

File: ci/azure/linux_script (hunk 1 above)
Comment (Vexu) [reply ci/azure/linux_script:64]: >     * This doesn't actually check any files other than src/main.zig

Solved with `find . -name  "*.zig" ! -path "*/zig-cache/*" ! -path "*/zig-out/*" | xargs -n1 release/bin/zig ast-check`.

>     * std lib AST checking is already happening with every stage2 test

Std files can be easily excluded in the command.

>     * It's an unrelated change to adding zig fmt checking

Sure, it did find a some bugs in stage2 and a lot of nonconforming code everywhere else so it might be better to do separately.  

>     * we may want to consider taking advantage of the thread pool and cache, which is already happening with std lib AST checking

It takes less than 15 seconds to check all ~900 files on my laptop on a debug build, I don't think performance is going to be that big of an issue.

File: ci/azure/linux_script [hunk 3]
```diff
@@ -61,6 +61,7 @@ make $JOBS install
 
 # look for formatting errors
 release/bin/zig fmt --check .. || (echo "Please run 'zig fmt' to fix the non-conforming files listed above." && false)
+find . -name  "*.zig" ! -path "*/zig-cache/*" ! -path "*/zig-out/*" ! -path "*/build/*" ! -wholename "*/special/compiler_rt.zig" ! -wholename "*/special/compiler_rt/atomics.zig"  | xargs -n 1 build/zig ast-check
```
Comment (Vexu) [inline ci/azure/linux_script:64]: I excluded some compiler_rt files since stage1 and stage2 don't agree on how `@export` works.

Comment (andrewrk) [reply ci/azure/linux_script:64]: instead of this, what about adding `--ast-check` to `zig fmt` since it's already doing the work of finding .zig files and excluding zig-cache?

File: lib/std/fmt/parse_float.zig [hunk 4]
```diff
@@ -34,7 +34,7 @@
 // - Only supports round-to-zero
 // - Does not handle denormals
 
-const std = @import("../std.zig");
+const std = @import("std");
```
Comment (g-w1) [inline lib/std/fmt/parse_float.zig:37]: Why was this changed? I think ../std.zig is more correct.

Comment (Vexu) [reply lib/std/fmt/parse_float.zig:37]: There is a pretty much even divide between `@import("std.zig")` and `@import("std")`, I changed this because I wanted to test this file specifically.

File: ci/azure/linux_script (hunk 3 above)
Comment (Vexu) [reply ci/azure/linux_script:64]: What should I do about the two compiler_rt files that do not currently pass the check due to the `@export` error?
```
error: the first @export parameter must be an identifier
        => @export(@import("compiler_rt/stack_probe.zig").zig_probe_stack, .{
                                                         ^
```

**Comment by Vexu** [comment]
```
$ hyperfine "zig-out/bin/zig fmt . --ast-check --check" --warmup 1
Benchmark #1: zig-out/bin/zig fmt . --ast-check --check
  Time (mean ± σ):      6.056 s ±  0.022 s    [User: 2.499 s, System: 3.552 s]
  Range (min … max):    6.022 s …  6.094 s    10 runs
```

Not too bad for ~860 files.

File: lib/std/fmt/parse_float.zig (hunk 4 above)
Comment (andrewrk) [reply lib/std/fmt/parse_float.zig:37]: Generally, std lib should import std.zig by file path, setting an example for other packages. For `zig test` you can use `--main-pkg-path` to test files that import with "..".
